### DISCUSSION TITLE
동시성 문제 해결

### DISCUSSION THREAD

**Discussion by woo-chang** [Busan-Dinosaur/Foodbowl-Backend#115] [개선해볼까요?]
현재 팔로우 같은 기능에서 같은 회원을 팔로우할 수 없다는 제약 조건이 존재합니다. 팔로우를 하기 전 해당 회원에 대한 팔로우 정보가 있는지 확인하는 로직이 존재하지만, 동시에 작업이 발생하는 경우 아래와 같은 상황이 만들어집니다.

![‎무제 ‎001](https://github.com/Busan-Dinosaur/Foodbowl-Backend/assets/78212016/392cc9c5-553b-4a02-81ff-d639798c773c)

실제 동시성 테스트를 하더라도 아래와 같은 예외를 확인할 수 있습니다.

<img width="1295" alt="스크린샷 2023-08-12 오후 12 42 14" src="https://github.com/Busan-Dinosaur/Foodbowl-Backend/assets/78212016/28473b56-b7a9-4b6f-ae1e-dddc3bdeafe1">

이에 대해 처리하지 않더라도 `Exception.class` 핸들러에 의해 처리되겠지만 클라이언트에서는 서버에서 알 수 없는 예외가 발생했다고 판단하게 됩니다. 커스텀 예외 메시지를 전달하기 위해서는 아래와 같은 코드를 작성해야 하는데 `DataIntegrityViolationException`을 `try~catch`로 잡아도 되는지에 대한 의문이 들었습니다. 여기서 잡게 된다면 DB와 소통하는 모든 코드에서 try~catch로 잡아야 하나 생각까지 들었기 때문입니다.

이 부분에 대해서 어떻게 하면 좋을지 얘기해보고 싶어요 :)

**Comment by Kim0914** [comment]
위와 같이 동시성 문제의 상황에서는, 클라이언트는 서버에서 알 수 없는 예외가 발생했다고 판단하는 것이 적절한 흐름이 아닐까요?!
클라이언트 입장에서 `동시성 문제에 의해서 예외가 발생했습니다.`의 정보가 필요한지 의문이 들어요.

동시성 문제를 고려한 상황까지 커스텀 예외를 도입해야하는가에 대해서 고민해보면 좋겠네요 :)

**Reply by woo-chang** [reply]
서버에서 발생하는 알 수 없는 예외 상황이 아닌,
`이미 팔로우중인 회원입니다.`라는 메시지를 클라이언트에게 전달할 수 있는 상황이라고 판단했습니다!

동시성 문제를 알리기보다는 **이미 팔로우중인 상황**을 알리고 싶었습니다.
해당 상황이 알 수 없는 서버 에러가 발생했다고 판단하기도 어려웠던 것 같아요 ..

이 관점에 대해서는 어떻게 생각하시는지 궁금합니다 🤓
